### ISSUE TITLE
Issue connecting to MySQL using SSL

### ISSUE THREAD

**Issue by LucasNN** [phpmyadmin/phpmyadmin#12146]
### Steps to reproduce
1) Configure the SSL connection in the config.inc.php

```
/**
 * MySQL hostname or IP address
 *
 * @global string $cfg['Servers'][$i]['host']
 */
$cfg['Servers'][$i]['host'] = '104.XXX.XXX.XX';

/**
 * MySQL port - leave blank for default port
 *
 * @global string $cfg['Servers'][$i]['port']
 */
$cfg['Servers'][$i]['port'] = '3306';

/**
 * MySQL extension
 *
 * @global string $cfg['Servers'][$i]['extension']
 */
$cfg['Servers'][$i]['extension'] = 'mysqli';

/**
 * Path to the socket - leave blank for default socket
 *
 * @global string $cfg['Servers'][$i]['socket']
 */
$cfg['Servers'][$i]['socket'] = '';

/**
 * Use SSL for connecting to MySQL server?
 *
 * @global boolean $cfg['Servers'][$i]['ssl']
 */
$cfg['Servers'][$i]['ssl'] = true;

/**
 * Path to the key file when using SSL for connecting to the MySQL server
 *
 * @global string $cfg['Servers'][$i]['ssl_key']
 */
$cfg['Servers'][$i]['ssl_key'] = '/opt/bitnami/apps/phpmyadmin/conf/certs/google-cloud-sql-client-key.pem';

/**
 * Path to the cert file when using SSL for connecting to the MySQL server
 *
 * @global string $cfg['Servers'][$i]['ssl_cert']
 */
$cfg['Servers'][$i]['ssl_cert'] = '/opt/bitnami/apps/phpmyadmin/conf/certs/google-cloud-sql-client-cert.pem';

/**
 * Path to the CA file when using SSL for connecting to the MySQL server
 *
 * @global string $cfg['Servers'][$i]['ssl_ca']
 */
$cfg['Servers'][$i]['ssl_ca'] = '/opt/bitnami/apps/phpmyadmin/conf/certs/google-cloud-sql-server-ca.pem';

/**
 * Directory containing trusted SSL CA certificates in PEM format
 *
 * @global string $cfg['Servers'][$i]['ssl_ca_path']
 */
$cfg['Servers'][$i]['ssl_ca_path'] = '/opt/bitnami/apps/phpmyadmin/conf/certs/';

/**
 * List of allowable ciphers for SSL connections to the MySQL server
 *
 * @global string $cfg['Servers'][$i]['ssl_ciphers']
 */
$cfg['Servers'][$i]['ssl_ciphers'] = null;
```

2) Try to sign in into the phpMyAdmin panel.

### Expected behaviour
Sign in into the phpMyAdmin panel normally.

### Actual behaviour
Errors:

```
Error

SQL query: Edit

SELECT * FROM information_schema.CHARACTER_SETS

MySQL said:

#2006 - MySQL server has gone away
```

```
Warning in ./libraries/dbi/DBIMysqli.php#252
 mysqli_query(): SSL operation failed with code 1. OpenSSL Error messages: error:0607A082:digital envelope routines:EVP_CIPHER_CTX_set_key_length:invalid key length error:0607A082:digital envelope routines:EVP_CIPHER_CTX_set_key_length:invalid key length

Backtrace

./libraries/dbi/DBIMysqli.php#252: mysqli_query(
object,
string 'SELECT * FROM information_schema.CHARACTER_SETS',
integer 0,
)
./libraries/DatabaseInterface.php#240: PMA\libraries\dbi\DBIMysqli->realQuery(
string 'SELECT * FROM information_schema.CHARACTER_SETS',
object,
integer 0,
)
./libraries/DatabaseInterface.php#89: PMA\libraries\DatabaseInterface->tryQuery(
string 'SELECT * FROM information_schema.CHARACTER_SETS',
NULL,
integer 0,
boolean true,
)
./libraries/mysql_charsets.inc.php#21: PMA\libraries\DatabaseInterface->query(string 'SELECT * FROM information_schema.CHARACTER_SETS')
./libraries/common.inc.php#905: include_once(./libraries/mysql_charsets.inc.php)
./index.php#13: require_once(./libraries/common.inc.php)
```

```
Warning in ./libraries/dbi/DBIMysqli.php#252
 mysqli_query(): MySQL server has gone away

Backtrace

./libraries/dbi/DBIMysqli.php#252: mysqli_query(
object,
string 'SELECT * FROM information_schema.CHARACTER_SETS',
integer 0,
)
./libraries/DatabaseInterface.php#240: PMA\libraries\dbi\DBIMysqli->realQuery(
string 'SELECT * FROM information_schema.CHARACTER_SETS',
object,
integer 0,
)
./libraries/DatabaseInterface.php#89: PMA\libraries\DatabaseInterface->tryQuery(
string 'SELECT * FROM information_schema.CHARACTER_SETS',
NULL,
integer 0,
boolean true,
)
./libraries/mysql_charsets.inc.php#21: PMA\libraries\DatabaseInterface->query(string 'SELECT * FROM information_schema.CHARACTER_SETS')
./libraries/common.inc.php#905: include_once(./libraries/mysql_charsets.inc.php)
./index.php#13: require_once(./libraries/common.inc.php)
```

```
Warning in ./libraries/dbi/DBIMysqli.php#252
 mysqli_query(): Error reading result set's header

Backtrace

./libraries/dbi/DBIMysqli.php#252: mysqli_query(
object,
string 'SELECT * FROM information_schema.CHARACTER_SETS',
integer 0,
)
./libraries/DatabaseInterface.php#240: PMA\libraries\dbi\DBIMysqli->realQuery(
string 'SELECT * FROM information_schema.CHARACTER_SETS',
object,
integer 0,
)
./libraries/DatabaseInterface.php#89: PMA\libraries\DatabaseInterface->tryQuery(
string 'SELECT * FROM information_schema.CHARACTER_SETS',
NULL,
integer 0,
boolean true,
)
./libraries/mysql_charsets.inc.php#21: PMA\libraries\DatabaseInterface->query(string 'SELECT * FROM information_schema.CHARACTER_SETS')
./libraries/common.inc.php#905: include_once(./libraries/mysql_charsets.inc.php)
./index.php#13: require_once(./libraries/common.inc.php)
```

### Server configuration
**Operating system**:
Google Cloud Compute Engine Bitnami LAMP Stack
Linux (Debian)

**Web server:**
Apache

**Database:**
MySQL 5.6.27 (Google Cloud SQL)

**PHP version:**
5.5.30

**phpMyAdmin version:**
4.6

### Client configuration
Chrome 49
OS X 10.8.5

**Comment by LucasNN** [comment]
#11553

**Comment by LucasNN** [comment]
![](http://s12.postimg.org/zero6pke5/sql_error_1.png)


![](http://s18.postimg.org/mgzfol4d5/sql_error_2.png)

**Comment by ibennetch** [comment]
One thing I noticed, I don't think `ssl_cipher` should be "null" &mdash; maybe blank (as in `='';`) or omitted entirely, but null definitely seems wrong here.

I'm not able to reproduce this problem on my local system. I wonder if there's something specific to Google Cloud SQL that's causing this. If your server is configured such that you have a user not requiring an SSL connection, I'm very curious about whether they're able to connect correctly.

Due to some MySQL issue, one thing you might try is to comment out all of the ssl lines *except* `$cfg['Servers'][$i]['ssl']` and `$cfg['Servers'][$i]['ssl_ca']`. I don't think it will help with this particular issue, but it would be interesting to try.

It's also worth trying the command line client to see if you experience the same trouble connecting from there. Most of the ssl options directly translate, so something like `mysql -u root -p -h 104.XXX.XXX.XX --ssl --ssl-cert=/opt/bitnami/apps/phpmyadmin/conf/certs/google-cloud-sql-client-cert.pem --ssl-ca=/opt/bitnami/apps/phpmyadmin/conf/certs/google-cloud-sql-server-ca.pem --ssl-capath=/opt/bitnami/apps/phpmyadmin/conf/certs/ --ssl-cipher=null` appears to be equivalent to your current phpMyAdmin configuration.

**Comment by LucasNN** [comment]
@ibennetch 

> One thing I noticed, I don't think ssl_cipher should be "null" — maybe blank (as in ='';) or omitted entirely, but null definitely seems wrong here.

Tried with ssl_cipher = ''; or omitted entirely, still not working.

> I'm not able to reproduce this problem on my local system. I wonder if there's something specific to Google Cloud SQL that's causing this. If your server is configured such that you have a user not requiring an SSL connection, I'm very curious about whether they're able to connect correctly.

Without SSL, phpMyAdmin works perfectly with Google Cloud SQL.

> Due to some MySQL issue, one thing you might try is to comment out all of the ssl lines except $cfg['Servers'][$i]['ssl'] and $cfg['Servers'][$i]['ssl_ca']. I don't think it will help with this particular issue, but it would be interesting to try.

Again, no success.

> It's also worth trying the command line client to see if you experience the same trouble connecting from there. Most of the ssl options directly translate, so something like mysql -u root -p -h 104.XXX.XXX.XX --ssl --ssl-cert=/opt/bitnami/apps/phpmyadmin/conf/certs/google-cloud-sql-client-cert.pem --ssl-ca=/opt/bitnami/apps/phpmyadmin/conf/certs/google-cloud-sql-server-ca.pem --ssl-capath=/opt/bitnami/apps/phpmyadmin/conf/certs/ --ssl-cipher=null appears to be equivalent to your current phpMyAdmin configuration.

Using this command I can connect successfully to Google Cloud SQL.

**Comment by nijel** [comment]
Can you try if changes in https://github.com/phpmyadmin/phpmyadmin/pull/12157 will help you?

**Comment by nijel** [comment]
I've just merged #12157 to QA_4_6 branch, it would be great if you could test if it will help in this case (I'm not sure, it's mostly cleanup to be sure we follow documentation).

**Comment by LucasNN** [comment]
Hey @nijel,

Thanks, but still not working.
=(

**Comment by LucasNN** [comment]
Tested everything and no success.

Does anyone know what could be?

**Comment by nijel** [comment]
Probably nobody here has experience with Google Cloud SQL. It might just generic problem with it, see eg. http://stackoverflow.com/questions/28881912/google-cloud-sql-mysql-server-has-gone-away

**Comment by LucasNN** [comment]
Hey @nijel,

I don't believe that the problem is in Google Cloud SQL.

Why?

Via Plaintext, I can connect regularly using:

1) mysql terminal command;
2) phpMyAdmin;
3) SQuirreL SQL.

Via SSL, I can connect regularly using:

1) mysql terminal command;
2) SQuirreL SQL.

So, I just have problem with phpMyAdmin, what is weird.

**Comment by jrchamp** [comment]
Google Cloud SQL has hostnames that won't match, so you need to turn off the verification:
`$cfg['Servers'][$i]['ssl_verify'] = false;`

**Comment by nijel** [comment]
Hmm, that however makes SSL pretty much useless...

**Comment by nijel** [comment]
Okay, I've set up Google Cloud SQL and did a bit of testing:

* Indeed the hostname validation fails, so ssl_verify = false is one of options.
* With my recent changes, ssl_verify doesn't work as expected on some PHP versions, I've fixed that.
* Once I connect I'm hitting #11553 
* Generally the PHP bindings suck at error reporting here - all they can do is to emit PHP warning with proper error

**Comment by nijel** [comment]
You can verify if SSL certification verification is the problem by using patch from https://github.com/phpmyadmin/phpmyadmin/pull/12193, that will probably not be the final patch as the errors are too noisy....

**Comment by nijel** [comment]
Pretty much this is bug of PHP and mysqlnd driver - it currently doesn't allow to disable host name validation without also disabling the certification validation completely. This has been initially discussed quite deeply at https://bugs.php.net/bug.php?id=68344 and I've new separate issue for that at https://bugs.php.net/bug.php?id=72048

**Comment by nijel** [comment]
I've added documentation to cover this topic: http://docs.phpmyadmin.net/en/latest/config.html#google-cloud-sql-with-ssl

**Comment by rjbrown99** [comment]
Would you mind updating the documentation to indicate that version 4.6.0.0 is the first release to support the ssl_verify option? The docs were helpful but took me some digging to figure out why it wasn't working for me. Thanks!

**Comment by nijel** [comment]
Good suggestion, added.

**Comment by WisdomEye** [comment]
For the issue about certificate validation, be sure have setup php correctly, I am using PHP7 on my test environment and I have setup on the file /usr/local/etc/php.ini:

```
...
[openssl]
openssl.cafile = "/usr/local/share/certs/ca-root-nss.crt"
...
```

(FreebSD 10.3 and phpmyadmin 4.6.1)

The next step problem (my guess is the core of problem) is about the cipher.

Testing using two servers I have Made:

MySQL Server on server1
MySQL Client on server 2

If I set on server1 in /usr/local/etc/my.cnf

```
...
[mysqld]
...
ssl_cipher=TLSv1.2
...
```

And on server2 in /usr/local/etc/my.cnf

```
[client]
...
ssl_cipher=TLSv1.2
...
```

(Have setup other variables to certificates too in both (ssl_ca, ssl_cert, ssl_key) and offcourse the test login was with require ssl setup.

Is possible connect using the mysql client without problem.
Same if change from TLSv1.2 to SSLv3.

But not work on phpmyadmin, with the errors:

TLSv1.2
Server does not answer

SSLv3
Error on handshaking

Other point I have noticed is, seems the extension is being ignored, have tried with both mysqli and mysql, but, error always popup saying its using mysqli only.

My guess the problem would be about connection syntax or bug in mysqli
